### PR TITLE
Documentation and tests

### PR DESCRIPTION
**Pull request by tevador**
This PR only contains changes in the documentation and tests. No actual RandomX code is affected.

One noteworthy change is completely rewritten design description and analysis (this is mainly to help auditors better understand the design).

### CODE REVIEW

File: doc/design.md [hunk 1]
```diff
@@ -1,158 +1,530 @@
...
-* Branch processing unit
-* Caches
-* Floating point unit
-* Arithmetic logic unit
-* Load-store unit
-* Memory controller
+### 1.2 The "Easy program problem"
 
-Modern CPUs also contain other circuits, such as PCI-E controllers, SATA, USB or Ethernet. Unfortunately, these cannot be used to perform any useful work.
+When a random program is generated, it can be executed only when it's favorable to do so.   This strategy is viable for two main reasons:
```
Comment (hyc) [inline doc/design.md]: I would say "When a random program is generated, one may choose to execute it only when it's favorable"

File: doc/design.md [hunk 2]
```diff
@@ -1,158 +1,530 @@
...
+
+#### 2.6.2 CBRANCH instruction
+
+RandomX therefore uses random branches with a jump probability of 1/256 and branch condition that depends on an integer register value. These branches will be predicted as "not taken" by the CPU. Such branches are "free" in most CPU designs unless they are taken. While this doesn't take advantage of the branch predictors, speculative designs will see a significant performance boost compared to non-speculative branch handling - see Appendix B for more information.
+
+The branching conditions and jump targets are chosen in such way that infinite loops in RandomX code are impossible because the register controlling the branch will never be modified in the repeated code block. Each CBRANCH instruction can jump up to twice in a row. Handling CBRANCH using predicated execution [[22](https://en.wikipedia.org/wiki/Predication_(computer_architecture))] is impractical because the branch is not take most of the time.
```
Comment (hyc) [inline doc/design.md]: s/not take /not taken /

File: doc/design.md [hunk 3]
```diff
@@ -1,158 +1,530 @@
...
+Additionally, integer instructions that read from a fixed address also use the whole "L3" Scratchpad (Table 5.1.4 of the Specification) because repetitive accesses will ensure that the cache line will be placed in the L1 cache of the CPU. This shows that the Scratchpad level doesn't always directly correspond to the same CPU cache level.
 
-Additionally, Scratchpad operations require write-read coherency, because when a write to L1 Scratchpad is in progress, a read has a 1/2048 chance of being from the same address. This is handled by the load-store unit (LSU) inside the CPU and requires every read to be checked against the addresses of all pending writes. Hardware without these coherency checks will produce >99% of invalid results.
+#### 2.8.2 Scratchpad writes
 
-While most writes to the Scratchpad are into L1 and L2, 144 bytes of data are written into L3 scratchpad per iteration, on average (64 bytes for integer registers, 64 bytes for floating point registers and 16 bytes by ISTORE instruction).
+There are two ways how Scratchpad is modified during VM execution:
```
Comment (hyc) [inline doc/design.md]: s/how/the/

File: doc/design.md [hunk 4]
```diff
@@ -1,158 +1,530 @@
...
+* The average amount of data read from memory per program iteration is: 39 * 8 + 128 + 64 = **504 bytes**.
+* The average mount of data written to memory per program iteration is: 16 * 8 + 128 = **256 bytes**.
+
+This is close to a 2:1 read/write ratio, which CPUs are optimized for.
+
+### 2.9 Dataset
+
+Since the Scratchpad is usually stored in the CPU cache, only Dataset accesses utilize the memory controllers.
+
+RandomX randomly reads from the Dataset once per program iteration (16384 times per hash result). Since the Dataset must be stored in DRAM, it provides a natural parallelization limit, because DRAM cannot do more than about 25 million random accesses per second per bank group. Each separately addressable bank group allows a throughput of around 1500 H/s.
```
Comment (hyc) [inline doc/design.md]: This text seems out of date with the earlier text about DDR3 vs DDR4 limits in the FAQ.

File: doc/design.md [hunk 5]
```diff
@@ -1,158 +1,530 @@
...
 
-#### AesGenerator
+## 3. Custom functions
 
-AesGenerator was designed for fastest possible generation of pseudorandom data. It takes advantage of hardware accelerated AES in modern CPUs. Only one AES round is performed per 16 bytes of output, which results in throughput exceeding 20 GB/s. The Scratchpad can be filled in under 100 μs. The AesGenerator state is initialized from the output of Blake2b.
+### 3.1 AesGenerator1R
 
-#### AesHash
+AesGenerator1R was designed for fastest possible generation of pseudorandom data to fill the Scratchpad. It takes advantage of hardware accelerated AES in modern CPUs. Only one AES round is performed per 16 bytes of output, which results in throughput exceeding 20 GB/s in most modern CPUs. While 1 AES round is not sufficient for a good distribution of random values, this is not an issue because the purpose is just to initialize the Scratchpad with random non-zero data.
```
Comment (hyc) [inline doc/design.md]: s/for fastest/for the fastest/

File: doc/design.md [hunk 6]
```diff
@@ -1,158 +1,530 @@
...
+
+### 3.2 AesGenerator4R
+
+AesGenerator4R uses 4 AES rounds to generate pseudorandom data for Program Buffer initialization. Since 2 AES rounds are sufficient for full avalanche of all input bits [[28](https://csrc.nist.gov/csrc/media/projects/cryptographic-standards-and-guidelines/documents/aes-development/rijndael-ammended.pdf)], AesGenerator4R provides an excellent output distribution while maintaining very good performance.
+
+The reversible nature of this generator is not an issue since the generator state is always initialized using the output of a non-reversible hashing function (Blake2b).
+
+### 3.3 AesHash1R
 
 AesHash was designed for fastest possible calculation of the Scratchpad fingerprint. It interprets the Scratchpad as a set of AES round keys, so it's equivalent to AES encryption with 32768 rounds. Two extra rounds are performed at the end to ensure avalanche of all Scratchpad bits in each lane. The output of the AesHash is fed into the Blake2b hashing function to calculate the final PoW hash.
```
Comment (hyc) [inline doc/design.md]: s/for fastest/for the fastest/

File: doc/design.md [hunk 7]
```diff
@@ -1,158 +1,530 @@
...
+
+Honest miner that doesn't reject any programs (`Q = 1`).
+
+#### Strategy II
+
+Miner that uses optimized custom hardware that cannot execute 25% of programs (`Q = 0.75`), but supported programs can be executed 50% faster.
+
+#### Strategy III
+
+Miner that can execute all programs, but rejects 25% of the slowest programs for the first program in the chain. This gives a 5% performance boost for the first program in the chain (this matches the runtime distribution from Appenix C).
```
Comment (hyc) [inline doc/design.md]: s/Appenix/Appendix/

File: doc/design.md (hunk 4 above)
Comment (tevador) [reply doc/design.md]: It should be still valid. DDR4 has multiple addressable banks per channel, which is what increases its performance. We should probably add some source to this paragraph, though.

**Comment by hyc** [comment]
One note from Trail of Bits, they were concerned with this in the Readme

> RandomX behaves like a keyed hashing function

since RandomX doesn't have the security properties of a keyed hash function. Mentioning this sort of thing nearly guarantees that some idiot will try to use it as a keyed hash, instead of as its intended purpose of a PoW algo.

**Comment by tevador** [comment]
I don't think that's an issue at all. The description of the repository clearly states "Proof of work algorithm" and it's stated several times in the readme as well. Also "behaves like" doesn't mean it is. Whoever decides to use RandomX should know what they are doing anyways.

**Pull request merged**